src/rewrites.rs

```diff
@@ -849,27 +1037,519 @@ impl MultiPatterns {
+                );
+            } else {
+                // Visited before. Check if in path_to_here
+                match path_to_here
+                    .iter()
+                    .enumerate()
+                    .find(|(i, (cid, n))| *cid == *child)
+                {
+                    Some((i, _)) => {
+                        let mut cycle: Vec<Mdl> =
+                            path_to_here[i..].iter().map(|(cid, n)| n.clone()).collect();
+                        cycle.push(node.clone());
+                        cycles.push(cycle);
+                    }
+                    None => (),
+                }
+            }
+        }
+    }
+}
+
+/// Update the blacklist_nodes in egraph.analysis with the new canonical EClass IDs
+fn update_blacklist(egraph: &mut EGraph<Mdl, TensorAnalysis>) {
+    egraph.analysis.blacklist_nodes = egraph
+        .analysis
+        .blacklist_nodes
+        .iter()
+        .map(|node| node.clone().map_children(|id| egraph.find(id)))
+        .collect();
+}
+
+/// Returns true if there will not be a cycle introduced by applying this rule.
+///
+/// Checking based on freshly collected descendents info. If any input node
+/// contains descendents that is any of the matched output class, then there can be a cycle
+/// created.
+///
+/// # Parameters
+///
+/// - `egraph`: egraph of interest
+/// - `input_subst`: substitution containing the input variables
+/// - `var_map_1`: keys of this map contains all the input variables in source pattern 1
+/// - `var_map_2`: keys of this map contains all the input variables in source pattern 2
+/// - `out_class_1`: Id of the matched eclass of the output of source pattern 1
+/// - `out_class_2`: Id of the matched eclass of the output of source pattern 2
+/// - `descendents`: Map from each eclass ID to its set of descendents. Constructed here.
+fn check_cycle(
+    egraph: &EGraph<Mdl, TensorAnalysis>,
+    input_subst: &Subst,
+    var_map_1: &HashMap<egg::Var, egg::Var>,
+    var_map_2: &HashMap<egg::Var, egg::Var>,
+    out_class_1: Id,
+    out_class_2: Id,
+    descendents: &mut HashMap<Id, HashSet<Id>>,
+) -> bool {
+    // Get all input eclass IDs
+    let input_ids: HashSet<Id> = var_map_1
+        .iter()
+        .chain(var_map_2.iter())
+        .map(|(var, _)| *input_subst.get(*var).unwrap())
+        .collect();
+    // Get a map from eclass IDs to eclass
+    let id_to_class: HashMap<Id, &EClass<Mdl, ValTnsr>> =
+        egraph.classes().map(|class| (class.id, class)).collect();
+    // Check descendents of the input eclasses
+    for id in input_ids.iter() {
+        get_descendents(
+            egraph,
+            *id,
+            &id_to_class,
+            /*check_blacklist=*/ false,
+            descendents,
+        );
+        let descendents_input = descendents.get(id).unwrap();
+        if descendents_input.contains(&out_class_1) || descendents_input.contains(&out_class_2) {
+            return false;
+        }
+    }
+    true
+}
```

Comment:
Same as above, you can use `return input_ids.iter().all(...)`

Reply:
Done